src/main/sensors/gyroanalyse.c

```diff
@@ -245,54 +269,91 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state)
+                binMax = lrintf(state->centerFreq[state->updateAxis] / fftResolution);
+            } else { // there was a max, find min
+                for (int i = binMax - 1; i > 1; i--) { // look for min below max
+                    dataMin = state->fftData[i];
+                    if (state->fftData[i - 1] > state->fftData[i]) { // up step below this one
+                        break;
+                    }
+                }
+                for (int i = binMax + 1; i < (FFT_BIN_COUNT - 1); i++) { // // look for min above max
+                    dataMinHi = state->fftData[i];
+                    if (state->fftData[i] < state->fftData[i + 1]) { // up step above this one
+                        break;
                     }
-
-                    fftSum += cubedData;
-                    // calculate weighted index starting at 1, not 0
-                    fftWeightedSum += cubedData * (i + 1);
                 }
             }
+            dataMin = fminf(dataMin, dataMinHi);
+
+            // accumulate fftSum and fftWeightedSum from peak bin, and shoulder bins either side of peak
+            float squaredData = state->fftData[binMax] * state->fftData[binMax];
+            float fftSum = squaredData;
+            float fftWeightedSum = squaredData * binMax;
+
+            // accumulate upper shoulder unless it would be FFT_BIN_COUNT
+            uint8_t shoulderBin = binMax + 1;
+            if (shoulderBin < FFT_BIN_COUNT) {
+                squaredData = state->fftData[shoulderBin] * state->fftData[shoulderBin];
+                fftSum += squaredData;
+                fftWeightedSum += squaredData * shoulderBin;
+            }
 
-            // get weighted center of relevant frequency range (this way we have a better resolution than 31.25Hz)
-            // if no peak, go to highest point to minimise delay
-            float centerFreq = dynNotchMaxCentreHz;
-            float fftMeanIndex = 0;
+            // accumulate lower shoulder unless lower shoulder would be bin 0 (DC)
+            if (binMax > 1) {
+                shoulderBin = binMax - 1;
+                squaredData = state->fftData[shoulderBin] * state->fftData[shoulderBin];
+                fftSum += squaredData;
+                fftWeightedSum += squaredData * shoulderBin;
+            }
 
+            // get centerFreq in Hz from weighted bins
+            float centerFreq = dynNotchMaxHz;
+            float fftMeanIndex = 0;
             if (fftSum > 0) {
-                // idx was shifted by 1 to start at 1, not 0
-                fftMeanIndex = (fftWeightedSum / fftSum) - 1;
-                // the index points at the center frequency of each bin so index 0 is actually 16.125Hz
-                centerFreq = constrain(fftMeanIndex * fftResolution, dynNotchMinHz, dynNotchMaxCentreHz);
+                fftMeanIndex = (fftWeightedSum / fftSum);
+                centerFreq = fftMeanIndex * fftResolution;
```

Comment:
Codacy found an issue: [Redundant initialization for 'centerFreq'. The initialized value is overwritten before it is read.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5547649)